File: 1/2020.10.18/106.Construct Binary Tree from Inorder and Postorder Traversal/python/solution.py
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if inorder:
            idx = inorder.index(postorder.pop())
            root = TreeNode(inorder[idx])
            root.right = self.buildTree(inorder[idx+1:], postorder)
            root.left = self.buildTree(inorder[:idx], postorder)
            return root

    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        def recur(low, high):
            if low > high:
                return None
            root = TreeNode(postorder.pop())
            mid = map_inorder[root.val]
            root.right = recur(mid + 1, high)
            root.left = recur(low, mid - 1)
            return root
        map_inorder = {}
        for i, val in enumerate(inorder):
            map_inorder[val] = i
        return recur(0, len(inorder) - 1)

File: 1/2020.10.18/106.Construct Binary Tree from Inorder and Postorder Traversal/java/solution.java
class Solution {

    Map<Integer, Integer> map = new HashMap<>();
    int rootIndex = 0;

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        rootIndex = postorder.length - 1;
        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i], i);
        }
        return recur(0, inorder.length - 1, inorder, postorder);
    }

    private TreeNode recur(int low, int high, int[] inorder, int[] postorder) {
        if (low > high) {
            return null;
        }
        TreeNode root = new TreeNode(postorder[rootIndex--]);
        int mid = map.get(root.val);
        root.right = recur(mid + 1, high, inorder, postorder);
        root.left = recur(low, mid - 1, inorder, postorder);
        return root;
    }
}
